azkaban-common/src/main/java/azkaban/project/FlowTrigger.java

```diff
@@ -0,0 +1,119 @@
+
+/**
+ * FlowTrigger is an immutable class which holds
+ * all the data and properties of a flow trigger.
+ */
+
+public class FlowTrigger {
+
+  private final List<Dependency> dependencies;
+  private final Props props; // trigger level props
+
+  private FlowTrigger(final Props props, final List<Dependency> dependencies) {
+    validateProps(props);
+    validateDependencies(dependencies);
+    this.props = new Props(props.getParent(), props);
+    this.dependencies = Collections.unmodifiableList(dependencies);
+  }
+
+  private void validateProps(final Props props) {
+    Preconditions.checkNotNull(props);
+    final String MISSING_REQUIRED_ERROR = "missing required param: %s";
+
+    final Set<String> requiredParam = ImmutableSet.of(Constants.TriggerProperties.SCHEDULE_TYPE,
+        Constants.TriggerProperties.SCHEDULE_VALUE,
+        Constants.TriggerProperties.SCHEDULE_MAX_WAIT_TIME);
+
+    for (final String param : requiredParam) {
+      Preconditions.checkArgument(props.containsKey(param), String.format(MISSING_REQUIRED_ERROR,
+          param));
+    }
+
+    Preconditions.checkArgument(props.size() == 3, String.format("invalid param found, allowed "
+        + "params: %s", StringUtils.join(requiredParam, ",")));
+
+    //todo chengren311: validate schedule type, value, and max wait time
+  }
+
+  private void validateDependencies(final List<Dependency> dependencies) {
+    Preconditions.checkNotNull(dependencies);
+    // at least one dependency for a trigger
+    Preconditions.checkArgument(!dependencies.isEmpty(), "no dependencies found");
+
+    // check uniqueness of dependency.name
+    Set<String> seen = Sets.newHashSet();
+    for (final Dependency dep : dependencies) {
+      Preconditions.checkArgument(seen.add(dep.getName()), String.format("duplicate dependency"
+          + ".name %s found, dependency.name should be unique", dep.getName()));
+    }
+
+    // check uniqueness of dependency config
+    seen = Sets.newHashSet();
+    for (final Dependency dep : dependencies) {
+      final Props props = dep.getProps();
+      props.removeLocal(Constants.DependencyProperties.DEPENDENCY_NAME);
+      Preconditions.checkArgument(seen.add(props.toString()), String.format("duplicate dependency"
+          + "config %s found, dependency config should be unique", dep.getName()));
+    }
+  }
+
+  @Override
+  public String toString() {
+    return "FlowTrigger{" +
+        "dependencies=" + this.dependencies +
+        ", props=" + this.props +
+        '}';
+  }
+
+  public List<Dependency> getDependencies() {
+    return this.dependencies;
+  }
+
+  public static class FlowTriggerBuilder {
+
+    private final List<Dependency> dependencies;
+    private final Props props; // trigger level props
+
+    public FlowTriggerBuilder(final Props props) {
+      this.props = props;
+      this.dependencies = Lists.newArrayList();
+    }
+
+    public FlowTriggerBuilder addDependency(final Dependency dep) {
```

Comment:
When converting yaml file to FlowTrigger, it would be convenient to have a setDependencies(final List<Dependency> dependencies) method. Do you mind if I add one later and change some of the code in the FlowTriggerBuilder? Or do you have any other suggestion?

Reply:
Personally I'm in favor of addDependency(final Dependency dep) than passing a list because there's no need to create a list outside the constructor. But it's not something big deal, if you feel setDependencies(final List dependencies)  is more convenient for you when parsing the yaml, then go ahead and change it.